lib/parsers/acorn-extensions.js

```diff
@@ -15,154 +15,180 @@ exports.enableTolerance = function (parser) {
+    node.declaration = parser.parseMaybeAssign();
   }
-  return this.finishNode(node, "ExportNamedDeclaration");
+  parser.semicolon();
+  return parser.finishNode(node, "ExportDefaultDeclaration");
 }
 
-// Calls the given callback with the state of the parser temporarily
-// advanced by calling this.nextToken() n times, then rolls the parser
-// back to its original state and returns whatever the callback returned.
-function withLookAhead(n, callback) {
-  const old = Object.assign(Object.create(null), this);
-  while (n-- > 0) this.nextToken();
-  try {
-    return callback(this);
-  } finally {
-    Object.assign(this, old);
+function parseExportFrom(parser, node) {
+  parser.expectContextual("from");
+  node.source = parser.type === tt.string ? parser.parseExprAtom() : null;
+  parser.semicolon();
+}
+
+function parseExportImportDefault(parser, node, exports) {
+  // export def from '...';
+  const specifier = parser.startNode();
+  specifier.exported = parser.parseIdent(true);
+
+  node.specifiers = [
+    parser.finishNode(specifier, "ExportDefaultSpecifier")
+  ];
+
+  if (parser.type === tt.comma &&
+      peekNextType(parser) === tt.star) {
+    // export def, * as ns from '...';
+    parser.next();
+    const specifier = parser.startNode();
+    parser.next();
+    parseExportNamespaceAs(parser, node, specifier, exports);
   }
+  // export def, * as ns [, { x, y as z }] from '...';
+  parseExportSpecifiersMaybe(parser, node);
+
+  parseExportFrom(parser, node);
+  return parser.finishNode(node, "ExportNamedDeclaration");
 }
 
-function peekNextType(parser) {
-  return parser.withLookAhead(1, () => parser.type);
+function parseExportImportSpecifiers(parser, node, exports) {
+  // export { x, y as z } [from '...'];
+  node.declaration = null;
+  node.specifiers = parser.parseExportSpecifiers(exports);
+
+  if (parser.isContextual("from")) {
+    parseExportFrom(parser, node, exports);
+  } else {
+    parser.semicolon();
+  }
+  return parser.finishNode(node, "ExportNamedDeclaration");
 }
 
-function isExportDefaultSpecifier() {
-  return this.type === tt.name &&
-    this.withLookAhead(1, isCommaOrFrom);
+function parseExportNamespace(parser, node, exports) {
+  const specifier = parser.startNode();
+
+  node.specifiers = [];
+  parser.next();
+
+  if (! parser.isContextual("as")) {
+    // export * from '...';
+    parseExportFrom(parser, node);
+    return parser.finishNode(node, "ExportAllDeclaration");
+  }
+  // export * as ns from '...';
+  parseExportNamespaceAs(parser, node, specifier, exports);
+  // export * as ns[, { x, y as z }] from '...';
+  parseExportSpecifiersMaybe(parser, node);
+
+  parseExportFrom(parser, node);
+  return parser.finishNode(node, "ExportNamedDeclaration");
 }
 
-function isCommaOrFrom(parser) {
-  return parser.type === tt.comma ||
-    (parser.type === tt.name &&
-     parser.value === "from");
+function parseExportNamespaceAs(parser, node, specifier, exports) {
```

Comment:
It's really only processing the "As" case though. Is there a more specific name than the generic parseExportNamespaceSpecifier?